crates/subspace-networking/src/utils.rs

```diff
@@ -95,3 +97,150 @@ pub(crate) fn convert_multiaddresses(addresses: Vec<Multiaddr>) -> Vec<PeerAddre
         })
         .collect()
 }
+
+// Semaphore like implementation that allows both shrinking and expanding
+// the max permits.
+#[derive(Clone, Debug)]
+pub(crate) struct ResizableSemaphore(Arc<SemShared>);
+
+// The permit.
+#[derive(Clone, Debug)]
+pub(crate) struct ResizableSemaphorePermit(Arc<SemShared>);
+
+// The state shared between the semaphore and the outstanding permits.
+#[derive(Debug)]
+struct SemShared {
+    // The tuple holds (current usage, current max capacity)
+    current: Mutex<SemState>,
```

Comment:
I'd call property `state` if it holds `SemState` data structure. `current` is much less clear.